### PR TITLE
Fixed idempotency with cpus parameter

### LINKED ISSUE #559: Container is recreated on every run when using the `cpus` directive
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When a container that has cgroup cpu limits applied by using the `cpus` directive, it is recreated each time the playbook runs even if there were no config changes made.

**Steps to reproduce the issue:**

1. Create an Ansible Playbook that should ensure a container is in the `started` state and is limited by using the `cpus` flag, for example the one shown below.

2. Run the playbook one first time to start the container

3. Rerun the playbook again without changing anything after the first run.

**Describe the results you received:**

The container was recreated, because the `cpu_period` flag was expected to be `0` but appears to be set to `100000` as default if a `--cpus` flag is provided. This results in a change detected by Ansible which then results in the container being recreated.

**Describe the results you expected:**

The container should be kept in the state it was and not be recreated or changed each time the playbook is rerun without any changes to it.

**Additional information you deem important (e.g. issue happens only occasionally):**

N/A

**Version of the `containers.podman` collection:**

```
containers.podman             1.10.1
```

**Output of `ansible --version`:**

```
ansible [core 2.14.2]
  config file = /home/lh/.ansible.cfg
  configured module search path = ['/home/lh/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.10/site-packages/ansible
  ansible collection location = /home/lh/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.9 (main, Dec 19 2022, 17:35:49) [GCC 12.2.0] (/usr/bin/python)
  jinja version = 3.1.2
  libyaml = True
```

**Output of `podman version`:**

```
Client:       Podman Engine
Version:      4.4.2
API Version:  4.4.2
Go Version:   go1.20.1
Git Commit:   74afe26887f814d1c39925a1624851ef3590e79c-dirty
Built:        Fri Feb 24 17:12:57 2023
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

``` yaml
host:
  arch: amd64
  buildahVersion: 1.29.0
  cgroupControllers:
  - cpu
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: /usr/bin/conmon is owned by conmon 1:2.1.6-1
    path: /usr/bin/conmon
    version: 'conmon version 2.1.6, commit: 158b5421dbac6bda96b1457955cf2e3c34af29bc'
  cpuUtilization:
    idlePercent: 91.77
    systemPercent: 1.5
    userPercent: 6.73
  cpus: 8
  distribution:
    distribution: arch
    version: unknown
  eventLogger: journald
  hostname: x1
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 10000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 10000
      size: 65536
  kernel: 6.2.1-arch1-1
  linkmode: dynamic
  logDriver: journald
  memFree: 824135680
  memTotal: 33350676480
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: /usr/bin/crun is owned by crun 1.8.1-1
    path: /usr/bin/crun
    version: |-
      crun version 1.8.1
      commit: f8a096be060b22ccd3d5f3ebe44108517fbf6c30
      rundir: /run/user/1000/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /etc/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: /usr/bin/slirp4netns is owned by slirp4netns 1.2.0-1
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.4
  swapFree: 8535142400
  swapTotal: 8589930496
  uptime: 19h 51m 33.00s (Approximately 0.79 days)
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries: {}
store:
  configFile: /home/lh/.config/containers/storage.conf
  containerStore:
    number: 2
    paused: 0
    running: 1
    stopped: 1
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/lh/.local/share/containers/storage
  graphRootAllocated: 1023667077120
  graphRootUsed: 572514779136
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 2
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/lh/.local/share/containers/storage/volumes
version:
  APIVersion: 4.4.2
  Built: 1677255177
  BuiltTime: Fri Feb 24 17:12:57 2023
  GitCommit: 74afe26887f814d1c39925a1624851ef3590e79c-dirty
  GoVersion: go1.20.1
  Os: linux
  OsArch: linux/amd64
  Version: 4.4.2
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
Name            : podman
Version         : 4.4.2-1
Description     : Tool and library for running OCI-based containers in pods
Architecture    : x86_64
URL             : https://github.com/containers/podman
Licenses        : Apache
Groups          : None
Provides        : None
Depends On      : catatonit  conmon  containers-common  crun  iptables  libdevmapper.so=1.02-64  libgpgme.so=11-64  libseccomp.so=2-64  slirp4netns
Optional Deps   : apparmor: for AppArmor support
                  btrfs-progs: support btrfs backend devices [installed]
                  cni-plugins: for an alternative container-network-stack implementation
                  fuse-overlayfs: for storage driver in rootless environment
                  podman-compose: for docker-compose compatibility
                  podman-docker: for Docker-compatible CLI
Required By     : None
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 79.03 MiB
Packager        : David Runge <dvzrv@archlinux.org>
Build Date      : Fri 24 Feb 2023 05:12:57 PM CET
Install Date    : Tue 28 Feb 2023 09:18:33 AM CET
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature
```

**Playbok you run with ansible (e.g. content of `playbook.yaml`):**

``` yaml
- name: container deployment
  hosts: localhost
  tasks:
    - name: start test container
      containers.podman.podman_container:
        name: "test-nginx-container"
        image: docker.io/nginx:latest
        publish:
          - 8080:80
        state: started
        memory: 256m
        cpus: 2.0
```

**Command line and output of ansible run with high verbosity**

`ansible-playbook --diff -vv playbook.yaml`

```
ansible-playbook [core 2.14.2]
  config file = /home/lh/.ansible.cfg
  configured module search path = ['/home/lh/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.10/site-packages/ansible
  ansible collection location = /home/lh/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible-playbook
  python version = 3.10.9 (main, Dec 19 2022, 17:35:49) [GCC 12.2.0] (/usr/bin/python)
  jinja version = 3.1.2
  libyaml = True
Using /home/lh/.ansible.cfg as config file
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not
match 'all'
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: playbook.yaml *******************************************************************************************
1 plays in playbook.yaml

PLAY [container deployment] ***************************************************************************************

TASK [Gathering Facts] ********************************************************************************************
task path: /home/lh/stuff/code/podman-ansible/playbook.yaml:1
ok: [localhost]

TASK [start test container] ***************************************************************************************
task path: /home/lh/stuff/code/podman-ansible/playbook.yaml:4
--- before
+++ after
@@ -1,2 +1,2 @@
-cpu_period - 100000
+cpu_period - 0
 cpus - 2.0

changed: [localhost] => {"actions": ["recreated test-nginx-container"], "changed": true, "container": {"AppArmorProfile": "", "Args": ["nginx", "-g", "daemon off;"], "BoundingCaps": ["CAP_CHOWN", "CAP_DAC_OVERRIDE", "CAP_FOWNER", "CAP_FSETID", "CAP_KILL", "CAP_NET_BIND_SERVICE", "CAP_SETFCAP", "CAP_SETGID", "CAP_SETPCAP", "CAP_SETUID"], "Config": {"Annotations": {"io.container.manager": "libpod", "org.opencontainers.image.stopSignal": "3"}, "AttachStderr": false, "AttachStdin": false, "AttachStdout": false, "Cmd": ["nginx", "-g", "daemon off;"], "CreateCommand": ["podman", "container", "run", "--name", "test-nginx-container", "--publish", "8080:80", "--memory", "256m", "--cpus", "2.0", "--detach=True", "docker.io/nginx:latest"], "Domainname": "", "Entrypoint": "/docker-entrypoint.sh", "Env": ["TERM=xterm", "container=podman", "NGINX_VERSION=1.23.3", "NJS_VERSION=0.7.9", "PKG_RELEASE=1~bullseye", "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", "HOME=/root", "HOSTNAME=9db10635ab13"], "HealthcheckOnFailureAction": "none", "Hostname": "9db10635ab13", "Image": "docker.io/library/nginx:latest", "Labels": {"maintainer": "NGINX Docker Maintainers <docker-maint@nginx.com>"}, "OnBuild": null, "OpenStdin": false, "Passwd": true, "StdinOnce": false, "StopSignal": 3, "StopTimeout": 10, "Timeout": 0, "Tty": false, "Umask": "0022", "User": "", "Volumes": null, "WorkingDir": "/", "sdNotifyMode": "container"}, "ConmonPidFile": "/run/user/1000/containers/overlay-containers/9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a/userdata/conmon.pid", "Created": "2023-03-01T11:06:38.552989477+01:00", "Dependencies": [], "Driver": "overlay", "EffectiveCaps": ["CAP_CHOWN", "CAP_DAC_OVERRIDE", "CAP_FOWNER", "CAP_FSETID", "CAP_KILL", "CAP_NET_BIND_SERVICE", "CAP_SETFCAP", "CAP_SETGID", "CAP_SETPCAP", "CAP_SETUID"], "ExecIDs": [], "GraphDriver": {"Data": {"LowerDir": "/home/lh/.local/share/containers/storage/overlay/94b65b4c2a149ea044f980aea763322202e038580cebfe8ffef4db325c870246/diff:/home/lh/.local/share/containers/storage/overlay/edeccf220855f936d6f8ccf86e05a89dd88451f0ff3f28c7f98e694b988a77e8/diff:/home/lh/.local/share/containers/storage/overlay/09dae7c0a5f24769002d90b9e20de329b8cb08dd7dafa7cac43a9b89d06c97a4/diff:/home/lh/.local/share/containers/storage/overlay/7090b39a83e99ee847e7a79b1bc6f013669cf98dbe52ab6f3b30972d6d7e2867/diff:/home/lh/.local/share/containers/storage/overlay/54e441944325a54657cfff3701ad28b2201a37d3a29a033731fbf2f20fb6338d/diff:/home/lh/.local/share/containers/storage/overlay/4695cdfb426a05673a100e69d2fe9810d9ab2b3dd88ead97c6a3627246d83815/diff", "MergedDir": "/home/lh/.local/share/containers/storage/overlay/03a9b1de2aeae8407cd24b213839712862c538d037a660162fd99f85291485c0/merged", "UpperDir": "/home/lh/.local/share/containers/storage/overlay/03a9b1de2aeae8407cd24b213839712862c538d037a660162fd99f85291485c0/diff", "WorkDir": "/home/lh/.local/share/containers/storage/overlay/03a9b1de2aeae8407cd24b213839712862c538d037a660162fd99f85291485c0/work"}, "Name": "overlay"}, "HostConfig": {"AutoRemove": false, "Binds": [], "BlkioDeviceReadBps": null, "BlkioDeviceReadIOps": null, "BlkioDeviceWriteBps": null, "BlkioDeviceWriteIOps": null, "BlkioWeight": 0, "BlkioWeightDevice": null, "CapAdd": [], "CapDrop": [], "Cgroup": "", "CgroupConf": null, "CgroupManager": "systemd", "CgroupMode": "private", "CgroupParent": "user.slice", "Cgroups": "default", "ConsoleSize": [0, 0], "ContainerIDFile": "", "CpuCount": 0, "CpuPercent": 0, "CpuPeriod": 100000, "CpuQuota": 200000, "CpuRealtimePeriod": 0, "CpuRealtimeRuntime": 0, "CpuShares": 0, "CpusetCpus": "", "CpusetMems": "", "Devices": [], "DiskQuota": 0, "Dns": [], "DnsOptions": [], "DnsSearch": [], "ExtraHosts": [], "GroupAdd": [], "IOMaximumBandwidth": 0, "IOMaximumIOps": 0, "IpcMode": "shareable", "Isolation": "", "KernelMemory": 0, "Links": null, "LogConfig": {"Config": null, "Path": "", "Size": "0B", "Tag": "", "Type": "journald"}, "Memory": 268435456, "MemoryReservation": 0, "MemorySwap": 536870912, "MemorySwappiness": -1, "NanoCpus": 2000000000, "NetworkMode": "slirp4netns", "OomKillDisable": false, "OomScoreAdj": 0, "PidMode": "private", "PidsLimit": 2048, "PortBindings": {"80/tcp": [{"HostIp": "", "HostPort": "8080"}]}, "Privileged": false, "PublishAllPorts": false, "ReadonlyRootfs": false, "RestartPolicy": {"MaximumRetryCount": 0, "Name": ""}, "Runtime": "oci", "SecurityOpt": [], "ShmSize": 65536000, "Tmpfs": {}, "UTSMode": "private", "Ulimits": [], "UsernsMode": "", "VolumeDriver": "", "VolumesFrom": null}, "HostnamePath": "/run/user/1000/containers/overlay-containers/9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a/userdata/hostname", "HostsPath": "/run/user/1000/containers/overlay-containers/9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a/userdata/hosts", "Id": "9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a", "Image": "3f8a00f137a0d2c8a2163a09901e28e2471999fde4efc2f9570b91f1c30acf94", "ImageDigest": "sha256:30bba6c267c24237c5b6f3c59a6e566365b80662f80935535ba3d6a694afe7e8", "ImageName": "docker.io/library/nginx:latest", "IsInfra": false, "IsService": false, "MountLabel": "", "Mounts": [], "Name": "test-nginx-container", "Namespace": "", "NetworkSettings": {"Bridge": "", "EndpointID": "", "Gateway": "", "GlobalIPv6Address": "", "GlobalIPv6PrefixLen": 0, "HairpinMode": false, "IPAddress": "", "IPPrefixLen": 0, "IPv6Gateway": "", "LinkLocalIPv6Address": "", "LinkLocalIPv6PrefixLen": 0, "MacAddress": "", "Ports": {"80/tcp": [{"HostIp": "", "HostPort": "8080"}]}, "SandboxID": "", "SandboxKey": "/run/user/1000/netns/netns-5d6f3bc3-f10b-4afb-655d-9bbf84524f2c"}, "OCIConfigPath": "/home/lh/.local/share/containers/storage/overlay-containers/9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a/userdata/config.json", "OCIRuntime": "crun", "Path": "/docker-entrypoint.sh", "PidFile": "/run/user/1000/containers/overlay-containers/9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a/userdata/pidfile", "Pod": "", "ProcessLabel": "", "ResolvConfPath": "/run/user/1000/containers/overlay-containers/9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a/userdata/resolv.conf", "RestartCount": 0, "Rootfs": "", "State": {"CgroupPath": "/user.slice/user-1000.slice/user@1000.service/user.slice/libpod-9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a.scope", "CheckpointedAt": "0001-01-01T00:00:00Z", "ConmonPid": 148121, "Dead": false, "Error": "", "ExitCode": 0, "FinishedAt": "0001-01-01T00:00:00Z", "Health": {"FailingStreak": 0, "Log": null, "Status": ""}, "OOMKilled": false, "OciVersion": "1.0.2-dev", "Paused": false, "Pid": 148123, "Restarting": false, "RestoredAt": "0001-01-01T00:00:00Z", "Running": true, "StartedAt": "2023-03-01T11:06:38.709218211+01:00", "Status": "running"}, "StaticDir": "/home/lh/.local/share/containers/storage/overlay-containers/9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a/userdata"}, "podman_actions": ["podman stop test-nginx-container", "podman rm -f test-nginx-container", "podman run --name test-nginx-container --publish 8080:80 --memory 256m --cpus 2.0 --detach=True docker.io/nginx:latest"], "podman_systemd": {"container-test-nginx-container": "# container-test-nginx-container.service\n# autogenerated by Podman 4.4.2\n# Wed Mar  1 11:06:38 CET 2023\n\n[Unit]\nDescription=Podman container-test-nginx-container.service\nDocumentation=man:podman-generate-systemd(1)\nWants=network-online.target\nAfter=network-online.target\nRequiresMountsFor=/run/user/1000/containers\n\n[Service]\nEnvironment=PODMAN_SYSTEMD_UNIT=%n\nRestart=on-failure\nTimeoutStopSec=70\nExecStart=/usr/bin/podman start test-nginx-container\nExecStop=/usr/bin/podman stop  \\\n\t-t 10 test-nginx-container\nExecStopPost=/usr/bin/podman stop  \\\n\t-t 10 test-nginx-container\nPIDFile=/run/user/1000/containers/overlay-containers/9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a/userdata/conmon.pid\nType=forking\n\n[Install]\nWantedBy=default.target\n"}, "stderr": "", "stderr_lines": [], "stdout": "9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a\n", "stdout_lines": ["9db10635ab1350d571abea73f30b63ffcb7c9b53b34f0b6d366dca84bdd1b61a"]}

PLAY RECAP ********************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

The described test case was run on a ThinkPad X1 Carbon Gen 9 running Arch Linux.

### PR DESCRIPTION
**Pull request by zarath**
Fixes #559

### CODE REVIEW

File: plugins/module_utils/podman/podman_container_lib.py [hunk 1]
```diff
@@ -1146,7 +1158,8 @@ def compose(p, h):
                     before.append(compose(port, h))
         after = self.params['publish'] or []
         if self.params['publish_all']:
-            image_ports = self.image_info.get('config', {}).get('exposedports', {})
+            image_ports = self.image_info.get(
```
Comment (sshnaidm) [inline plugins/module_utils/podman/podman_container_lib.py:1161]: Let's not change not related code (we have 120 chars lines in Ansible allowed)

**Review by sshnaidm — CHANGES_REQUESTED** [review]
Thanks for the contribution! A few questions are inside though.

File: plugins/module_utils/podman/podman_container_lib.py [hunk 2]
```diff
@@ -826,9 +830,16 @@ def diffparam_conmon_pidfile(self):
 
     def diffparam_cpu_period(self):
         before = self.info['hostconfig']['cpuperiod']
-        after = self.params['cpu_period']
+        # if cpu_period left to default keep settings
```
[the remark this answers is not in the document]
Comment (zarath) [reply plugins/module_utils/podman/podman_container_lib.py:833]: One can argue if it is more expected not to change an undefined parameter in a second run with different settings, or reset to default.
Personally I would prefer first.
I mean, if I intentionally use a parameter, I would require to set it back if I want to have a change.

It's a little more complex to ignore before as you have to view on both related parameters..

[the remark this answers is not in the document]
Comment (zarath) [reply plugins/module_utils/podman/podman_container_lib.py:833]: They are related.
If I don't add the new parameter, the fix will not work.
https://docs.podman.io/en/latest/markdown/podman-run.1.html
> --cpus Number of CPUs. The default is 0.0 which means no limit. This is shorthand for --cpu-period and --cpu-quota, therefore the option cannot be specified with --cpu-period or --cpu-quota.
> 

To reach idempotency all cases have to be solved as info returns both representations, no matter if any parameter was given. As soon as you put one in the other appears in the delta.

[the remark this answers is not in the document]
Comment (sshnaidm) [reply plugins/module_utils/podman/podman_container_lib.py:833]: Sorry, I struggle to understand how exactly they are related. If you remove all with `cpu_quota` from your patch, it still fixes the idempotency issue, you can check it with simple task running twice:
```yaml
    - name: Create test container
      containers.podman.podman_container:
        name: test
        image: alpine:3.16
        state: present
        command: sleep 1h
        cpus: 2.0
```

[the remark this answers is not in the document]
Comment (sshnaidm) [reply plugins/module_utils/podman/podman_container_lib.py:833]: I suggested we merge the fix for current issue, and then merge addition of `cpu_quota`, that's it.

[the remark this answers is not in the document]
Comment (zarath) [reply plugins/module_utils/podman/podman_container_lib.py:829]: done

**Review by sshnaidm — APPROVED** [review]
Awesome, thanks!

**Pull request merged**